File: python/dmp_base/CanonicalSystem.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Mon Oct 30 19:00:00 2017

@author: gsutanto
"""

import numpy as np
import os
import sys
sys.path.append(os.path.join(os.path.dirname(__file__), '../dmp_param/'))
sys.path.append(os.path.join(os.path.dirname(__file__), '../dmp_coupling/base/'))
from TauSystem import *
from CanonicalCoupling import *

class CanonicalSystem:
    'Base class for canonical systems of DMPs.'
    
    def __init__(self, name, tau_system, canonical_couplers_list):
        self.name = name
        self.tau_sys = tau_system
        self.canonical_couplers_list = canonical_couplers_list
        self.is_started = False
    
    def isValid(self):
        assert (self.tau_sys != None), "TauSystem tau_sys does NOT exist!"
        assert (self.tau_sys.isValid() == True), "TauSystem tau_sys is invalid!"
        return True
    
    def getCouplingTerm(self):
        accumulated_cc = 0.0
        for canonical_coupler_idx in range(len(self.canonical_couplers_list)):
            cc = self.canonical_couplers_list[canonical_coupler_idx].getValue()
            assert (math.isnan(cc) == False), 'cc['+str(canonical_coupler_idx)+'] is NaN!'
            accumulated_cc = accumulated_cc + cc
        return accumulated_cc

File: src/dmp_base/CanonicalSystem.cpp
#include "dmp/dmp_base/CanonicalSystem.h"

namespace dmp {

CanonicalSystem::CanonicalSystem()
    : tau_sys(NULL),
      canonical_coupling(NULL),
      is_started(false),
      rt_assertor(NULL) {}

CanonicalSystem::CanonicalSystem(
    TauSystem* tau_system, RealTimeAssertor* real_time_assertor,
    std::vector<CanonicalCoupling*>* canonical_couplers)
    : tau_sys(tau_system),
      canonical_coupling(canonical_couplers),
      is_started(false),
      rt_assertor(real_time_assertor) {}

bool CanonicalSystem::isValid() {
  if (rt_assert(tau_sys != NULL) == false) {
    return false;
  }
  if (rt_assert(tau_sys->isValid()) == false) {
    return false;
  }
  return true;
}

bool CanonicalSystem::getCouplingTerm(double& accumulated_cc) {
  accumulated_cc = 0.0;

  if (canonical_coupling) {
    for (uint i = 0; i < canonical_coupling->size(); ++i) {
      if ((*canonical_coupling)[i] == NULL) {
        continue;
      }

      double cc = 0.0;
      if (rt_assert((*canonical_coupling)[i]->getValue(cc)) == false) {
        return false;
      }
      if (rt_assert(std::isnan(cc) == 0) == false) {
        return false;
      }

      accumulated_cc += cc;
    }
  }

  return true;
}

TauSystem* CanonicalSystem::getTauSystemPointer() { return tau_sys; }

CanonicalSystem::~CanonicalSystem() {}
}  // namespace dmp
